src/server/transport/socketio.ts

```diff
@@ -126,7 +129,11 @@ export class SocketIO {
             TransportAPI(matchID, socket, this.clientInfo, this.roomInfo),
             this.auth
           );
-          await master.onUpdate(action, stateID, matchID, playerID);
+
+          this.createMatchQueue(matchID);
+          await this.perMatchQueue[matchID].add(() =>
```

Comment:
Is it actually possible to access a map entry like this without using `get`? I wonder if you could implement a `getMatchQueue` method instead of `createMatchQueue` which will make sure the match queue exists and then return it for use.

Reply:
Sure, I'll work on it.

Reply:
That's the problem on using AsyncInMemory class on Master tests: the tests only work on the Master level, not on the socket level where the queue is. We have to implement a separate test for the socket using master with AsyncInMemory database.

Reply:
@delucis , here are the news. I've made some practical tests today to test my implementation of PQueue for onUpdate messages with InMemory database on my game. It didn't work well, some parts of the state where overwritten and the clients became out of sync. Although, testing it with FlatFile worked perfectly which made me think FlatFile works better for the ignoreStaleStateID options because of that concurrency management you mentioned. The queue system was almost irrelevant, I need to test more because without the queue I saw just a few problems.

Reply:
I've made a few more tests with PQueue and FlatFile and I had no problems. After that I put a 300 milliseconds interval to PQueue for Sync-typed storages, like InMemory, and it worked fine on my app. Also, I've found this article where it shows an example of testing SocketIO with Jest.

> https://medium.com/@tozwierz/testing-socket-io-with-jest-on-backend-node-js-f71f7ec7010f

I'll try it tomorrow, that's the only way I see for testing the update queuing on socket level.

Reply:
> It didn't work well, some parts of the state where overwritten and the clients became out of sync.

That’s good to know — looks like we must be missing something important! I don’t have much time in the next few days to look into this, but let me know how you get on with the tests.

Reply:
Thanks, @delucis , no rush. I'm facing some difficulties with the tests, there're some things I don't know how they work. Here is the test I wrote. First, I tried with just the app server but I had a problem where ctx was always undefined. Then I tried to use the API with the lobby client to create a match and authenticate the players but I got that `fetch`, from lobby client, is not defined.

Do you know how can I create and start a match without using lobby API but using the SocketIO server and client?

```javascript
describe('simultaneous moves', () => {
  const game = {
    name: 'test',
    setup: () => {
      const G = {
        players: {
          '0': {
            cards: ['card3'],
          },
          '1': {
            cards: [],
          },
        },
        cards: ['card0', 'card1', 'card2'],
        discardedCards: [],
      };
      return G;
    },
    playerView: PlayerView.STRIP_SECRETS,
    turn: {
      activePlayers: { currentPlayer: { stage: 'A' } },
      stages: {
        A: {
          moves: {
            A: (G, ctx: Ctx) => {
              const card = G.players[ctx.playerID].cards.shift();
              G.discardedCards.push(card);
            },
            B: {
              move: (G, ctx: Ctx) => {
                const card = G.cards.pop();
                G.players[ctx.playerID].cards.push(card);
              },
              ignoreStaleStateID: true,
            },
          },
        },
      },
    },
  };
  let server: ReturnType<typeof Server> | null;
  let runningServer: { appServer: KoaServer; apiServer?: KoaServer } | null;
  let clientConfig;
  let httpAppAddr;
  let client0: _ClientImpl<any>;
  let client1: _ClientImpl<any>;
  let lobbyClient: LobbyClient;

  beforeEach(async () => {
    server = Server({ games: [game] });
    runningServer = await server.run(undefined);
    httpAppAddr = runningServer.appServer.address();

    // lobbyClient = new LobbyClient({ server: `http://[${httpAppAddr.address}]:${httpAppAddr.port}` });

    clientConfig = {
      game: game,
      multiplayer: SocketIOClient({ server: `http://[${httpAppAddr.address}]:${httpAppAddr.port}` }),
      numPlayers: 2,
    };

    // const match = await lobbyClient.createMatch('test', { numPlayers: 2 });

    /*const p0Auth = await lobbyClient.joinMatch('test', match.matchID, {
      playerID: '0',
      playerName: 'Player 0',
    });
    const p1Auth = await lobbyClient.joinMatch('test', match.matchID, {
      playerID: '1',
      playerName: 'Player 1',
    });*/

    client0 = Client({
      ...clientConfig,
      playerID: '0',
      // credentials: p0Auth.playerCredentials,
    });
    client1 = Client({
      ...clientConfig,
      playerID: '1',
      credentials: p1Auth.playerCredentials,
    });

    client0.start();
    client1.start();

    client0.updateMatchID(match.matchID);
    client1.updateMatchID(match.matchID);

    // const state = { G: {}, ctx: { phase: '' }, plugins: {} };
    // const filteredMetadata = [];
    // client0.store.dispatch(sync({ state, filteredMetadata } as SyncInfo));
    // client1.store.dispatch(sync({ state, filteredMetadata } as SyncInfo));

    const state = { restore: true };
    const log = ['0', '1'];
    const action = sync(({ state, log } as unknown) as SyncInfo);
    client0.store.dispatch(action);
    client1.store.dispatch(action);
  });

  afterEach(() => {
    if (server && runningServer) {
      const { apiServer, appServer } = runningServer;
      server.kill({ apiServer, appServer });
    }

    if (client0 && client0.stop) client0.stop();
    if (client1 && client1.stop) client1.stop();
  });

  test('two clients playing', async () => {
    expect(server).not.toBeUndefined();
    expect(runningServer.appServer).not.toBeUndefined();
    expect(runningServer.apiServer).toBeUndefined();

    // test: simultaneous moves
    client0.getInitialState();
    client1.getInitialState();

    client0.moves.A();
    client0.events.setActivePlayers({ all: 'A' });

    client1.moves.B();
    client0.moves.B();

    const stateP0 = client0.getState();
    const stateP1 = client1.getState();

    expect(stateP0.G).toMatchObject({
      players: {
        '0': {
          cards: ['card1'],
        },
      },
      cards: ['card0'],
      discardedCards: ['card3'],
    });
    expect(stateP1.G).toMatchObject({
      players: {
        '1': {
          cards: ['card2'],
        },
      },
      cards: ['card0'],
      discardedCards: ['card3'],
    });
  });
});

```

Reply:
I can’t see here where the storage is instantiated, but could you use it directly to create state, e.g.:

```js
const db = new Storage();

const server = Server({ db });

db.createMatch('id', {
  initialState: InitializeGame({ /* ... */ }),
  metadata: createMetadata({ /* ... */ }),
});
```

`InitializeGame` is exported from [`core/initialize`](https://github.com/boardgameio/boardgame.io/blob/master/src/core/initialize.ts#L17) and `createMetadata` from [`server/util`](https://github.com/boardgameio/boardgame.io/blob/master/src/server/util.ts#L6) if you decide to use them, but you could also use manual values.

Otherwise, when `client.start()` is called, connecting the transport, a `SYNC` action should be dispatched to the master, which will create the match automatically.

Reply:
Thanks, @delucis. About sending a SYNC action, how can I do this from the clients? Am I missing something here?

Reply:
It should happen automatically. Basically when you call `client.start()`, it calls `transport.connect()`. In the transport implementations, a sync action is dispatched as part of the connect method.

Reply:
That's the problem, the `socket.on('sync', () => { })` is not being called on the client. Looks like the client is not connected to the server or the server didn't respond.

Reply:
From the logs, it looks like the server starts after the client move. How?  :confused: 

```
simultaneous moves › two clients playing

    TypeError: Cannot read property 'phase' of undefined

      167 | 
      168 |   function GetPhase(ctx: { phase: string }): PhaseConfig {
    > 169 |     return ctx.phase ? phaseMap[ctx.phase] : phaseMap[''];
          |                ^
      170 |   }
      171 | 
      172 |   function OnMove(s) {

      at GetPhase (src/core/flow.ts:169:16)
      at Object.GetMove (src/core/flow.ts:632:18)
      at src/core/reducer.ts:190:38
      at dispatch (node_modules/redux/lib/redux.js:218:22)
      at src/client/client.ts:226:22
      at src/client/client.ts:280:22
      at Object.dispatch (src/client/client.ts:295:22)
      at Object.dispatchers.<computed> [as A] (src/client/client.ts:85:13)
      at Object.<anonymous> (src/server/transport/socket-multiplayer.test.ts:143:19)

  console.log src/core/logger.ts:14
    INFO: App serving on 34173...
```

Reply:
Sad news. I've tried a lot of things. That approach is not going to work under a test environment, [see more here](https://stackoverflow.com/questions/15509231/unit-testing-node-js-and-websockets-socket-io). I will try again mocking the server SocketIO and send fake moves.

Reply:
Thanks for your persistence! Hopefully I’ll have more time soon to actually read through the changes. Mocking sounds like a good idea.

Reply:
> (By the way, noticed the “Skipping failing test just to push he code.” commit. You can push without running tests by doing `git push --no-verify` if that helps in the future.)

Thanks, I forgot about that.  :smile:

Reply:
Now I see: Master methods onSync and onUpdate are mocked.

Reply:
@delucis Finally!  :smile:  I've managed to write the unit tests for the actions queue on socket level. I've also took liberty to add playerID and action type on Master's error messages, it help me to follow the actions being sent to the server and possible errors.

Reply:
Nice! 😄

I’ll try to get a review in soon.